Podcast feature

Ivan12273 opened the pull request:
Context: Adding a new feature for listen to podcast from iTunes API.

Hello! I am working on the issue [#966](https://github.com/nukeop/nuclear/issues/966) I added the next changes:

- **rest/iTunesPodcast.ts:** four functions for obtaining the information from the iTunes API.
- **stream/iTunesPodcastPlugin.ts:** search and getAlternateStream (I needed to do two request for each one, one for obtaining the id of the podcast and the other for the episode, the iTunes API has its limitations for searching)
- **meta/iTunespodcast.ts:** for obtaining the details of the podcast and the episodes. I did it in artist in order to test what I have, I was trying to do it in album but I noticed that the album tab works different that the artist tab.

I solved my problem I mentioned with the id.
I was thinking of creating a new podcast tab or re-use the album tab, but still not sure how the album components work, I would appreciate if someone can help guide me through the parts that I need to modify to add a new tab as album :)

nukeop commented:
> I was thinking of creating a new podcast tab or re-use the album tab, but still not sure how the album components work, I would appreciate if someone can help guide me through the parts that I need to modify to add a new tab as album :)

I think it would be good to just use the same album grid as albums use.

Search results are provided by this container: https://github.com/nukeop/nuclear/blob/master/packages/app/app/containers/SearchResultsContainer/index.js
And they're passed to this component: https://github.com/nukeop/nuclear/blob/master/packages/app/app/components/SearchResults/index.js

If you format the results correctly, you will be able to add an item under `panes()`, using `renderPane`.
If you're able, please also add a test for this here: https://github.com/nukeop/nuclear/tree/master/packages/app/app/containers/SearchResultsContainer

Ivan12273 commented:
Hello, I made new changes in:

- /components/SearchResults/index.js
- /containers/SearchResultsContainer/index.js

And other files in order to add a new tab for podcast, I re-use the album grid and other album functions for display the podcasts.

Also I added a new abstract function in metaProvider.

![image](https://user-images.githubusercontent.com/42753018/122627379-9df82080-d074-11eb-94e6-df4bbefd1adf.png)

![image](https://user-images.githubusercontent.com/42753018/122627418-cf70ec00-d074-11eb-9da1-69c0f94048c2.png)

It worked fine, are these changes correct?

HaiDang666 commented on packages/app/app/components/AlbumView/index.tsx:
```diff
@@ -138,7 +138,7 @@ export const AlbumView: React.FC<AlbumViewProps> = ({
                     <Icon name='hashtag' />
                   </th>
-                  <th className={styles.left}>Song</th>
+                  <th className={styles.left}>Tracks</th>
```
can you use locale in stead of this hard string

Ivan12273 replied:
Sorry, what did you mean with locale?

Ivan12273 replied:
I changed it to how it was originally

nukeop replied:
These need to be translated according to the selected language, using `react-i18next`.

HaiDang666 replied:
you can see in the same file that:
```nodejs
const { t } = useTranslation('album');
...
<PopupButton
    onClick={addAlbumToDownloads}
    ariaLabel={t('download')}
    icon='download'
    label={t('download')}
  />
```
and the `download` key is defined in `packages/i18n/src/locales/en.json` as:
```js
{
  ...
  album: {
     ...
     download: 'Download'
  }
}
```

Ivan12273 replied:
Hi! I changed it again using `react-i18next` I was only able to translate to Spanish, the others are in English :)

HaiDang666 replied:
That good 😌

HaiDang666 commented on packages/core/src/plugins/meta/audius.ts:
```diff
@@ -128,10 +129,13 @@ class AudiusMetaProvider extends MetaProvider {
     throw new Error('Method not implemented.');
   }
+
+  searchForPodcast(): Promise<SearchResultsPodcast[]> {
```
@nukeop is it possible to make an interface for `MetaProvider` to implement default method to return empty value. So we don't need to do like this in very meta provider when we come ups with new method.

nukeop replied:
@Ivan12273 You can add this method to this abstract class with a default implementation: https://github.com/nukeop/nuclear/blob/6a99b3f85042f170de840d3e9b9381d8cd5fd84b/packages/core/src/plugins/metaProvider.ts

This way you can remove it from child providers.

Ivan12273 replied:
I added the method with the default implementation and removed it from the others classes. I needed to add` // eslint-disable-next-line @typescript-eslint/no-unused-vars` for the unused var :)

nukeop commented:
I agree with @HaiDang666 's comments. Very good work too.

HaiDang666 commented:
I found test snapshot fail and this
<img width="949" alt="Screen Shot 2021-06-20 at 16 39 39" src="https://user-images.githubusercontent.com/12780733/122669216-788b2400-d1e6-11eb-809e-03c41bc6edeb.png">

Ivan12273 commented:
> I found test snapshot fail and this
> <img alt="Screen Shot 2021-06-20 at 16 39 39" width="949" src="https://user-images.githubusercontent.com/12780733/122669216-788b2400-d1e6-11eb-809e-03c41bc6edeb.png">
> 
> <img alt="Screen Shot 2021-06-20 at 16 43 50" width="425" src="https://user-images.githubusercontent.com/12780733/122669274-bdaf5600-d1e6-11eb-93c6-7cb6a43fed49.png">

Hi, I corrected it, it was because I changed the text haha.

![image](https://user-images.githubusercontent.com/42753018/122676044-c29de680-d1a1-11eb-85e3-db5f744ea040.png)

The first one is because of the file name, in my local project I have it different, I already changed it in the repo too :)

HaiDang666 reviewed (approved):
Good work, just wait for @nukeop.

nukeop commented:
Thanks, this is a really good contribution.
